xclim/indices/_helpers.py

```diff
@@ -0,0 +1,266 @@
+    # The following equation is not explictely stated in the reference but it can easily be derived.
+    if stat == "integral":
+        return 2 * (
+            h_s * np.sin(declination) * np.sin(lat)
+            + np.cos(declination) * np.cos(lat) * np.sin(h_s)
+        )
+    if stat == "average":
+        return (
+            np.sin(declination) * np.sin(lat)
+            + np.cos(declination) * np.cos(lat) * np.sin(h_s) / h_s
+        )
+
+
+def extraterrestrial_solar_radiation(
+    times: xr.DataArray,
+    lat: xr.DataArray,
+    solar_constant: str = "1361 W m-2",
+    method="spencer",
+) -> xr.DataArray:
+    """Extraterrestrial solar radiation
+
+    This is the daily energy received on a surface parallel to the ground at the mean
+    distance of the earth to the sun. It neglects the effect of the atmosphere. Computation
+    is based on [Kalogirou14]_ and the default solar constant is taken from [Matthes17]_.
+
+    Parameters
+    ----------
+    times: xr.DataArray
+            Daily datetime data. This function makes no sense with data of other frequency.
+    latitude : xr.DataArray
+        Latitude.
+
+    Returns
+    -------
+    Extraterrestrial solar radiation, [J m-2 d-1]
+
+    Reference
+    ---------
+    .. [Matthes17] Matthes, K. et al. (2017). Solar forcing for CMIP6 (v3.2). Geoscientific Model Development, 10(6), 2247–2302. https://doi.org/10.5194/gmd-10-2247-2017
+    .. [Kalogirou14] Kalogirou, S. A. (2014). Chapter 2 — Environmental Characteristics. In S. A. Kalogirou (Ed.), Solar Energy Engineering (Second Edition) (pp. 51–123). Academic Press. https://doi.org/10.1016/B978-0-12-397270-5.00002-9
+    """
+    da = ((datetime_to_decimal_year(times) % 1) * 2 * np.pi).assign_attrs(units="rad")
+    dr = eccentricity_correction_factor(da, method=method)
+    ds = solar_declination(da, method=method)
+    gsc = convert_units_to(solar_constant, "J m-2 d-1")
+    rad_to_day = 1 / (2 * np.pi)  # convert radians of the "day circle" to day
+    return (
+        gsc * rad_to_day * cosine_of_solar_zenith_angle(ds, lat, stat="integral") * dr
+    ).assign_attrs(units="J m-2 d-1")
+
+
+def day_lengths(
+    dates: xr.DataArray,
+    lat: xr.DataArray,
+    start_date: xr.DataArray | DayOfYearStr | None = None,
+    end_date: xr.DataArray | DayOfYearStr | None = None,
+    freq: str = "YS",
+    method: str = "spencer",
+) -> xr.DataArray:
+    r"""Day-lengths according to latitude and day of year.
+
+    See :py:func:`solar_declination` for the approximation used to compute the solar
+    declination angle. Based on [Kalogirou14]_.
+
+    Parameters
+    ----------
+    dates: xr.DataArray
+    lat: xarray.DataArray
+      Latitude coordinate.
+    start_date: xarray.DataArray or DayOfYearStr, optional
+      Start date to consider for calculating mean day lengths. Default: None.
+    end_date: xarray.DataArray or DayOfYearStr, optional
+      End date to consider for calculating mean day lengths. Default: None.
```

Comment:
Is that necessary? If that's a common use case, an option would be to add an example in the docstring.

Reply:
Yes, the start and end dates are needed for agroclimate indices. An example of an application: doi:10.1016/j.agrformet.2003.06.001

Reply:
@Zeitsperre I thought about it and maybe this is not a use case common enough to be implemented directly in the `day_lengths` function? My next commit moves the aggregation to the BEDD and Huglin indices.